### ISSUE TITLE
Introduction to linear mixed models

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#3816]
Introduction to linear mixed models<br>
<br>
http://bit.ly/2wqWO59<br>
<br>
About 0 Minutes<br>
<br>
<div><div>

<section>
<div>
<div>
<div>
<div>
<p>By <span>Gabriela K Hajduk</span> on <span>March 15, 2017</span></p>
<hr>
<div>
<center>
<img alt="Img" src="https://ourcodingclub.github.io/img/tutheadermixed.png">
</center>
</div>
<p>This workshop is aimed at people new to mixed modeling and as such, it doesn’t cover all the nuances of mixed models, but hopefully serves as a starting point when it comes to both the concepts and the code syntax in <code>R</code>. There are no equations used to keep it beginner friendly.</p>
<p><strong>Acknowledgements:</strong> First of all, thanks where thanks are due. This tutorial has been built on the tutorial written by <a href="https://twitter.com/ldbailey255" target="_blank">Liam Bailey</a>, who has been kind enough to let me use chunks of his script, as well as some of the data. Having this backbone of code made my life much, much easier, so thanks Liam, you are a star! The seemingly excessive waffling is mine.</p>
<p>If you are familiar with linear models, aware of their shortcomings and happy with their fitting, then you should be able to very quickly get through the first five sections below. I am however including them for the sake of completeness and in an attempt to cater to a broader audience.</p>
<p>Similarly, I include quite a bit of explanatory text: you might choose to just skim it for now and go through the “coding bits” of the tutorial. But it will be here to help you along when you start using mixed models with your own data and you need a bit more context.</p>
<p>To get all you need for this session, <strong>go to <a href="https://github.com/ourcodingclub/CC-Linear-mixed-models" target="_blank">the repository for this tutorial</a>, click on <code>Clone/Download/Download ZIP</code> to download the files and then unzip the folder. Alternatively, fork the repository to your own Github account, clone the repository on your computer and start a version-controlled project in RStudio. For more details on how to do this, please check out our <a href="https://ourcodingclub.github.io/2017/02/27/git.html" target="_blank"> Intro to Github for version control</a> tutorial.</strong></p>
<p>Alternatively, you can grab the <strong>R script</strong> <a href="http://gkhajduk.d.pr/FG8/2bCpZQuj" target="_blank">here</a> and the <strong>data</strong> from <a href="http://gkhajduk.d.pr/9GPn/3nbbPoK6" target="_blank">here</a>. I might update this tutorial in the future and if I do, the latest version will be <a href="https://gkhajduk.github.io/2017-03-09-mixed-models/" target="_blank">on my website</a>.</p>
<h2>Tutorial Sections:</h2>
<h3><a href="#one"> 1. What is mixed effects modelling and why does it matter?</a></h3>
<h3><a href="#two"> 2. Explore the data</a></h3>
<h3><a href="#three"> 3. Fit all data in one analysis</a></h3>
<h3><a href="#four"> 4. Run multiple analyses</a></h3>
<h3><a href="#five"> 5. Modify the current model</a></h3>
<h3><a href="#six"> 6. Mixed effects models</a></h3>
<h4><a href="#FERE"> – Fixed and Random effects</a></h4>
<h4><a href="#first"> – Let’s fit our first mixed model</a></h4>
<h4><a href="#types"> – Types of random effects</a></h4>
<h5><a href="#crossed"> Crossed random effects</a></h5>
<h5><a href="#nested"> Nested random effects</a></h5>
<h5><a href="#implicit"> Implicit vs. explicit nesting</a></h5>
<h4><a href="#second"> – Our second mixed model</a></h4>
<h4><a href="#presenting"> – Presenting your model results</a></h4>
<h5><a href="#tables"> Tables</a></h5>
<h5><a href="#dot"> Dot-and-Whisker plots</a></h5>
<h5><a href="#processing"> Further processing</a></h5>
<h4><a href="#extra"> – EXTRA: P-values and model selection</a></h4>
<h5><a href="#fixedstr"> Fixed effects structure</a></h5>
<h5><a href="#randomstr"> Random effects structure</a></h5>
<h5><a href="#selection"> The entire model selection</a></h5>
<h3><a href="#end"> 7. THE END </a></h3>
<h3>What is mixed effects modelling and why does it matter?</h3>
<p>Ecological and biological data are often complex and messy. We can have different <strong>grouping factors</strong> like populations, species, sites we collect the data at etc. <strong>Sample sizes</strong> might leave something to be desired too, especially if we are trying to fit complicated models with <strong>many parameters</strong>. On top of that, our data points might <strong>not be truly independent</strong>. For instance, we might be using quadrats within our sites to collect the data (and so there is structure to our data: quadrats are nested within the sites).</p>
<p>This is why <strong>mixed models</strong> were developed, to deal with such messy data and to allow us to use all our data, even when we have low sample sizes, structured data and many covariates to fit. Oh, and on top of all that, mixed models allow us to save degrees of freedom compared to running standard linear models! Sounds good, doesn’t it?</p>
<p>We will cover only linear mixed models here, but if you are trying to “extend” your linear model, fear not: there are generalised linear mixed effects models out there, too.</p>
<h3>Explore the data</h3>
<p>We are going to focus on a fictional study system, dragons, so that we don’t have to get too distracted with the specifics of this example (and so I don’t throw too much biology/ecology at those of you who come from different fields). Imagine that we decided to train dragons and so we went out into the mountains and collected data on dragon intelligence (<code>testScore</code>) as a prerequisite. We sampled individuals with a range of body lengths across three sites in eight different mountain ranges. Start by loading the data and having a look at them.</p>
<div><div><pre><code><span>load</span><span>(</span><span>"dragons.RData"</span><span>)</span><span>
</span><span>head</span><span>(</span><span>dragons</span><span>)</span><span>
</span></code></pre></div></div>
<p>Let’s say we want to know how the body length of the dragons affects their test scores.</p>
<p>You don’t need to worry about the distribution of your <strong>explanatory</strong> variables. Have a look at the distribution of the response variable:</p>
<div><div><pre><code><span>hist</span><span>(</span><span>dragons</span><span>$</span><span>testScore</span><span>)</span><span> </span><span># seems close to a normal distribution - good!</span><span>
</span></code></pre></div></div>
<center><img alt="Img" src="https://ourcodingclub.github.io/img/mm-1.png"></center>
<p>It is good practice to <strong>standardise</strong> your explanatory variables before proceeding so that they have a mean of zero and standard deviation of one. It ensures that the estimated coefficients are all on the same scale, making it easier to compare effect sizes. You can use <code>scale()</code> to do that:</p>
<div><div><pre><code><span>dragons</span><span>$</span><span>bodyLength2</span><span> </span><span>&lt;-</span><span> </span><span>scale</span><span>(</span><span>dragons</span><span>$</span><span>bodyLength</span><span>)</span><span>
</span></code></pre></div></div>
<p><code>scale()</code> centers the data (the column mean is subtracted from the values in the column) and then scales it (the centered column values are divided by the column’s standard deviation).</p>
<p>Back to our question: is the test score affected by body length?</p>
<h3>Fit all data in one analysis</h3>
<p>One way to analyse this data would be to fit a linear model to all our data, ignoring the sites and the mountain ranges for now.</p>
<p>Fit the model with <code>testScore</code> as the response and <code>bodyLength2</code> as the predictor and have a look at the output:</p>
<div><div><pre><code><span>basic.lm</span><span> </span><span>&lt;-</span><span> </span><span>lm</span><span>(</span><span>testScore</span><span> </span><span>~</span><span> </span><span>bodyLength2</span><span>,</span><span> </span><span>data</span><span> </span><span>=</span><span> </span><span>dragons</span><span>)</span><span>
</span><span>summary</span><span>(</span><span>basic.lm</span><span>)</span><span>
</span></code></pre></div></div>
<p>Let’s plot the data with ggplot2.</p>
<div><div><pre><code><span>library</span><span>(</span><span>ggplot2</span><span>)</span><span> </span><span># load the package</span><span>
</span><span>(</span><span>prelim_plot</span><span> </span><span>&lt;-</span><span> </span><span>ggplot</span><span>(</span><span>dragons</span><span>,</span><span> </span><span>aes</span><span>(</span><span>x</span><span> </span><span>=</span><span> </span><span>bodyLength</span><span>,</span><span> </span><span>y</span><span> </span><span>=</span><span> </span><span>testScore</span><span>))</span><span> </span><span>+</span><span>
</span><span>geom_point</span><span>()</span><span> </span><span>+</span><span>
</span><span>geom_smooth</span><span>(</span><span>method</span><span> </span><span>=</span><span> </span><span>"lm"</span><span>))</span><span>
</span></code></pre></div></div>
<p>Note that putting your entire ggplot code in brackets () creates the graph and then shows it in the plot viewer. If you don’t have the brackets, you’ve only created the object, but haven’t visualised it. You would then have to call the object such that it will be displayed by just typing <code>prelim_plot</code> after you’ve created the “prelim_plot” object.</p>
<center><img alt="Img" src="https://ourcodingclub.github.io/img/mm-2.png"></center>
<p>Okay, so both from the linear model and from the plot, it seems like bigger dragons do better in our intelligence test. That seems a bit odd: size shouldn’t really affect the test scores.</p>
<p>But… are the assumptions met?</p>
<p>Plot the residuals: the red line should be nearly flat, like the dashed grey line:</p>
<div><div><pre><code><span>plot</span><span>(</span><span>basic.lm</span><span>,</span><span> </span><span>which</span><span> </span><span>=</span><span> </span><span>1</span><span>)</span><span> </span><span># not perfect... </span><span>
</span><span>## but since this is a fictional example we will go with it</span><span>
</span><span>## for your own data be careful:</span><span>
</span><span>## the bigger the sample size, the less of a trend you'd expect to see</span><span>
</span></code></pre></div></div>
<center><img alt="Img" src="https://ourcodingclub.github.io/img/mm-3.png"></center>
<p>Have a quick look at the qqplot too: points should ideally fall onto the diagonal dashed line:</p>
<div><div><pre><code><span>plot</span><span>(</span><span>basic.lm</span><span>,</span><span> </span><span>which</span><span> </span><span>=</span><span> </span><span>2</span><span>)</span><span> </span><span># a bit off at the extremes, but that's often the case; again doesn't look too bad</span><span>
</span></code></pre></div></div>
<center><img alt="Img" src="https://ourcodingclub.github.io/img/mm-4.png"></center>
<p>However, what about observation independence? Are our data independent?</p>
<p>We collected multiple samples from eight mountain ranges. It’s perfectly plausible that the data from within each mountain range are more similar to each other than the data from different mountain ranges: they are correlated.</p>
<p>Have a look at the data to see if above is true:</p>
<div><div><pre><code><span>boxplot</span><span>(</span><span>testScore</span><span> </span><span>~</span><span> </span><span>mountainRange</span><span>,</span><span> </span><span>data</span><span> </span><span>=</span><span> </span><span>dragons</span><span>)</span><span> </span><span># certainly looks like something is going on here</span><span>
</span></code></pre></div></div>
<center><img alt="Img" src="https://ourcodingclub.github.io/img/mm-5.png"></center>
<p>We could also plot it and colour points by mountain range:</p>
<div><div><pre><code><span>(</span><span>colour_plot</span><span> </span><span>&lt;-</span><span> </span><span>ggplot</span><span>(</span><span>dragons</span><span>,</span><span> </span><span>aes</span><span>(</span><span>x</span><span> </span><span>=</span><span> </span><span>bodyLength</span><span>,</span><span> </span><span>y</span><span> </span><span>=</span><span> </span><span>testScore</span><span>,</span><span> </span><span>colour</span><span> </span><span>=</span><span> </span><span>mountainRange</span><span>))</span><span> </span><span>+</span><span>
</span><span>geom_point</span><span>(</span><span>size</span><span> </span><span>=</span><span> </span><span>2</span><span>)</span><span> </span><span>+</span><span>
</span><span>theme_classic</span><span>()</span><span> </span><span>+</span><span>
</span><span>theme</span><span>(</span><span>legend.position</span><span> </span><span>=</span><span> </span><span>"none"</span><span>))</span><span>
</span></code></pre></div></div>
<center><img alt="Img" src="https://ourcodingclub.github.io/img/mm-6.png"></center>
<p>From the above plots, it looks like our mountain ranges vary both in the dragon body length and in their test scores. This confirms that our observations from within each of the ranges <strong>aren’t independent</strong>. We can’t ignore that.</p>
<p>So what do we do?</p>
<h3>Run multiple analyses</h3>
<p>We could run many separate analyses and fit a regression for each of the mountain ranges.</p>
<p>Lets have a quick look at the data split by mountain range. We use the <code>facet_wrap</code> to do that:</p>
<div><div><pre><code><span>(</span><span>split_plot</span><span> </span><span>&lt;-</span><span> </span><span>ggplot</span><span>(</span><span>aes</span><span>(</span><span>bodyLength</span><span>,</span><span> </span><span>testScore</span><span>),</span><span> </span><span>data</span><span> </span><span>=</span><span> </span><span>dragons</span><span>)</span><span> </span><span>+</span><span> </span><span>geom_point</span><span>()</span><span> </span><span>+</span><span> </span><span>facet_wrap</span><span>(</span><span>~</span><span> </span><span>mountainRange</span><span>)</span><span> </span><span>+</span><span> </span><span>xlab</span><span>(</span><span>"length"</span><span>)</span><span> </span><span>+</span><span> </span><span>ylab</span><span>(</span><span>"test score"</span><span>))</span><span>
</span></code></pre></div></div>
<center><img alt="Img" src="https://ourcodingclub.github.io/img/mm-7.png"></center>
<p>That’s eight analyses. Oh wait, we also have different sites, which similarly to mountain ranges aren’t independent… So we could run an analysis for each site in each range separately.</p>
<p>To do the above, we would have to estimate a slope and intercept parameter for each regression. That’s two parameters, three sites and eight mountain ranges, which means <strong>48 parameter estimates</strong> (2 x 3 x 8 = 48)! Moreover, the sample size for each analysis would be only 20.</p>
<p>This presents problems: not only are we <strong>hugely decreasing our sample size</strong>, but we are also <strong>increasing chances of a Type I Error (where you falsely reject the null hypothesis) by carrying out multiple comparisons</strong>. Not ideal!</p>
<h3>Modify the current model</h3>
<p>We want to use all the data, but account for the data coming from different mountain ranges (let’s put sites on hold for a second to make things simpler).</p>
<p>Add mountain range as a fixed effect to our <code>basic.lm</code></p>
<div><div><pre><code><span>mountain.lm</span><span> </span><span>&lt;-</span><span> </span><span>lm</span><span>(</span><span>testScore</span><span> </span><span>~</span><span> </span><span>bodyLength2</span><span> </span><span>+</span><span> </span><span>mountainRange</span><span>,</span><span> </span><span>data</span><span> </span><span>=</span><span> </span><span>dragons</span><span>)</span><span>
</span><span>summary</span><span>(</span><span>mountain.lm</span><span>)</span><span>
</span></code></pre></div></div>
<p>Now body length is <u>not</u> significant. But let’s think about what we are doing here for a second. The above model is estimating the difference in test scores between the mountain ranges - we can see all of them in the model output returned by <code>summary()</code>. But we are not interested in that, we just want to know whether body length affects test scores and we want to simply <strong>control for the variation</strong> coming from mountain ranges.</p>
<p>This is what we refer to as <strong>“random factors”</strong> and so we arrive at mixed effects models. Ta-daa!</p>
<h3>Mixed effects models</h3>
<p>A Mixed model is a good choice here: it will allow us to <strong>use all the data we have</strong> (higher sample size) and <strong>account for the correlations between data</strong> coming from the sites and mountain ranges. We will also <strong>estimate fewer parameters</strong> and <strong>avoid problems with multiple comparisons</strong> that we would encounter while using separate regressions.</p>
<p>We are going to work in <code>lme4</code>, so load the package (or use <code>install.packages</code> if you don’t have <code>lme4</code> on your computer).</p>
<h4>Fixed and Random effects</h4>
<p>Let’s talk a little about the <strong>fixed and random effects</strong> first. The literature isn’t clear on the exact definitions of those so I’m going to give you an “introductory” explanation. See links in the further reading below if you want to know more.</p>
<p>In some cases, the same variable could be considered either a random or a fixed effect (and sometimes even both at the same time!) so you have to think not only about your data, but also <strong>about the questions you are asking</strong> and construct your models accordingly.</p>
<p>In broad terms, <strong>fixed effects</strong> are variables that we expect will have an effect on the dependent/response variable. In our case, we are interested in making conclusions about how dragon body length impacts the dragon’s test score. So body length is a fixed effect and test score is the dependent variable.</p>
<p>On the other hand, <strong>random effects</strong> (or random factors - as they will be <strong>categorical</strong>, you can’t force R to treat a continuous variable as a random effect) are usually <strong>grouping factors</strong> for which we are trying to control. A lot of the time we are not specifically interested in their impact on the response variable. Additionally, the data for our random effect is just <strong>a sample of all the possibilities</strong>. Keep in mind that <em>random</em> doesn’t have much to do with <em>mathematical randomness</em>. Yes, it’s confusing. Just think about them as the <em>grouping</em> variables for now.</p>
<p>Strictly speaking it’s all about making our models better <strong>and getting better estimates</strong>.</p>
<p>In this particular case, we are looking to control for the effects of mountain range. We haven’t sampled all the mountain ranges in the world (we have eight) so our data are just a sample of all the existing mountain ranges. We are not really interested in the effect of each specific mountain range on the test score, but we know that the test scores from within the ranges might be correlated so we want to control for that.</p>
<p>If we specifically chose eight particular mountain ranges <em>a priori</em> and we were interested in those ranges and wanted to make predictions about them, then mountain range would be fitted as a fixed effect.</p>
<p><strong>NOTE:</strong> Generally you want your random effect to have <strong>more than five levels</strong>. So, for instance, if we wanted to control for the effects of dragon’s sex on intelligence, we would fit sex (a two level factor: male or female) <strong>as a fixed, not random, effect</strong>.</p>
<h5>
<strong>So the big question is:</strong> <em>what are you trying to do? What are you trying to make predictions about? What is just variation (a.k.a “noise”) that you need to control for?</em>
</h5>
<h4>Further reading for the keen:</h4>
<ul>
<li>
<p><a href="https://dynamicecology.wordpress.com/2015/11/04/is-it-a-fixed-or-random-effect/" target="_blank">Is it a fixed or random effect?</a> A useful way to think about fixed <em>vs</em>. random effects is in terms of partitioning the variation and estimating random effects with <strong>partial pooling</strong>. The description <a href="http://stats.stackexchange.com/questions/4700/what-is-the-difference-between-fixed-effect-random-effect-and-mixed-effect-mode" target="_blank">here</a> is the most accessible one I could find for now and you can find more opinions in the comments under the previous link too (search for <em>pooling</em> and <em>shrinkage</em> too if you are very keen).</p>
</li>
<li>
<p><a href="https://dynamicecology.wordpress.com/2015/02/05/how-many-terms-in-your-model-before-statistical-machismo/" target="_blank">How many terms? On model complexity</a></p>
</li>
</ul>
<p><a href="https://dynamicecology.wordpress.com/2014/12/02/why-are-your-statistical-models-more-complex-these-days/" target="_blank">More on model complexity</a></p>
<ul>
<li>Have a look at some of the fixed and random effects definitions gathered by Gelman in <a href="http://www.stat.columbia.edu/~gelman/research/published/AOS259.pdf" target="_blank">this paper</a> (you can also find them <a href="http://stats.stackexchange.com/questions/4700/what-is-the-difference-between-fixed-effect-random-effect-and-mixed-effect-mode/4702#4702" target="_blank">here</a> if you can’t access the paper).</li>
</ul>
<h3>Let’s fit our first mixed model</h3>
<p>Alright! Still with me? We have a response variable, the test score and we are attempting to <strong>explain part of the variation</strong> in test score through fitting body length as a fixed effect. But the response variable has some <strong>residual variation</strong> (<em>i.e.</em> unexplained variation) associated with mountain ranges. By using random effects, we are modeling that unexplained variation through <strong>variance</strong>.</p>
<p>[Sidenote: If you are confused between variation and variance: <strong>variation</strong> is a generic word, similar to dispersion or variability; <strong>variance</strong> is a particular measure of variation; it quantifies the dispersion, if you wish.]</p>
<p>Note that <strong>our question changes slightly here</strong>: while we still want to know whether there is an association between dragon’s body length and the test score, we want to know if that association exists <strong><em>after</em></strong> controlling for the variation in mountain ranges.</p>
<p>We will fit the random effect using <code>(1|variableName)</code>:</p>
<div><div><pre><code><span>mixed.lmer</span><span> </span><span>&lt;-</span><span> </span><span>lmer</span><span>(</span><span>testScore</span><span> </span><span>~</span><span> </span><span>bodyLength2</span><span> </span><span>+</span><span> </span><span>(</span><span>1</span><span>|</span><span>mountainRange</span><span>),</span><span> </span><span>data</span><span> </span><span>=</span><span> </span><span>dragons</span><span>)</span><span>
</span><span>summary</span><span>(</span><span>mixed.lmer</span><span>)</span><span>
</span></code></pre></div></div>
<p>Once we account for the mountain ranges, it’s obvious that dragon body length doesn’t actually explain the differences in the test scores.</p>
<p>Keep in mind that the random effect of the mountain range is <strong>meant to capture all the influences of mountain ranges on dragon test scores</strong> - whether we observed those influences explicitly or not, whether those influences are big or small <em>etc</em>. It could be many, many teeny-tiny influences that, when combined, affect the test scores and that’s what we are hoping to control for.</p>
<p>As always, it’s good practice to have a look at the plots to check our assumptions:</p>
<div><div><pre><code><span>plot</span><span>(</span><span>mixed.lmer</span><span>)</span><span> </span><span># looks alright, no paterns evident</span><span>
</span></code></pre></div></div>
<center><img alt="Img" src="https://ourcodingclub.github.io/img/mm-8.png"></center>
<p>and “<code>qqplot</code>”:</p>
<div><div><pre><code><span>qqnorm</span><span>(</span><span>resid</span><span>(</span><span>mixed.lmer</span><span>))</span><span>
</span><span>qqline</span><span>(</span><span>resid</span><span>(</span><span>mixed.lmer</span><span>))</span><span> </span><span># points fall nicely onto the line - good!</span><span>
</span></code></pre></div></div>
<center><img alt="Img" src="https://ourcodingclub.github.io/img/mm-9.png"></center>
<p>Let’s go back to the summary and look at our results again.</p>
<p>We can see the variance for <code>mountainRange = 339.7</code>. Mountain ranges are clearly important: they explain a lot of variation. How do we know that? We can take the variance for the <code>mountainRange</code> and divide it by the total variance:</p>
<div><div><pre><code><span>339.7</span><span>/</span><span>(</span><span>339.7</span><span> </span><span>+</span><span> </span><span>223.8</span><span>)</span><span> </span><span># ~60 %</span><span>
</span></code></pre></div></div>
<p>So the differences between mountain ranges explain ~60% of the variance. Do keep in mind that’s 60% of variance “left over” after the variance explained by our fixed effects.</p>
<h4>Types of random effects</h4>
<p>Before we go any further, let’s review the syntax above and chat about crossed and nested random effects. It’s useful to get those clear in your head.</p>
<p><strong>Reminder</strong>: a factor is just any categorical independent variable.</p>
<p>Above, we used <code>(1|mountainRange)</code> to fit our random effect. Whatever is on the right side of <code>|</code> operator is a factor and referred to as a “grouping factor” for the term.</p>
<p><strong>Random effects (factors) can be crossed or nested</strong> - it depends on the relationship between the variables. Let’s have a look.</p>
<h5>Crossed random effects</h5>
<p>Be careful with the nomenclature. There are <strong>“hierarchical linear models”</strong> (HLMs) or <strong>“multilevel models”</strong> out there, but while all HLMs are mixed models, <strong>not all mixed models are hierarchical</strong>. That’s because you can have <strong>crossed (or partially crossed) random factors</strong> that do not represent levels in a hierarchy.</p>
<p>Think for instance about our study where you monitor dragons (subject) across different mountain ranges (context) and imagine that we collect <strong>multiple observations per dragon</strong> (we give it the test multiple times - risking <strong>pseudoreplication</strong>). Since our dragons can fly, it’s easy to imagine that <strong>we might observe the same dragon across different mountain ranges</strong>, but also that we might not see all the dragons visiting all of the mountain ranges. Therefore, we can potentially observe every dragon in every mountain range (<strong>crossed</strong>) or at least observe some dragons across some of the mountain ranges (<strong>partially crossed</strong>). We would then fit the identity of the dragon and mountain range as (partially) crossed random effects.</p>
<h4>Nested random effects</h4>
<p>If this sounds confusing, not to worry - <code>lme4</code> handles partially and fully crossed factors well. They don’t have to be hierarchical or “multilevel” by design. However, <strong>the same model specification can be used to represent both (partially) crossed or nested factors</strong> so you can’t use the model’s specification to tell you what’s going on with the random factors: you have to look at the structure of the factors in the data. To make things easier for yourself, code your data properly and <strong>avoid implicit nesting</strong>. Not sure what implicit nesting is? Read on.</p>
<h4>Implicit <em>vs</em>. explicit nesting</h4>
<p>To tackle this, let’s look at another aspect of our study: we collected the data on dragons not only across multiple mountain ranges, but also across several sites within those mountain ranges. If you don’t remember have another look at the data:</p>
<div><div><pre><code><span>head</span><span>(</span><span>dragons</span><span>)</span><span> </span><span># we have site and mountainRange</span><span>
</span><span>str</span><span>(</span><span>dragons</span><span>)</span><span> </span><span># we took samples from three sites per mountain range and eight mountain ranges in total</span><span>
</span></code></pre></div></div>
<p>Just like we did with the mountain ranges, we have to assume that data collected within our sites might be <strong>correlated</strong> and so we should include sites as <strong>an additional random effect</strong> in our model.</p>
<p>Our site variable is a three-level factor, with sites called <code>a</code>, <code>b</code> and <code>c</code>. The nesting of the site within the mountain range is <strong>implicit</strong> - our sites are meaningless without being assigned to specific mountain ranges, i.e. there is nothing linking site <code>b</code> of the <code>Bavarian</code> mountain range with site <code>b</code> of the <code>Central</code> mountain range. To avoid future confusion we should create a new variable that is <strong>explicitly nested</strong>. Let’s call it <code>sample</code>:</p>
<div><div><pre><code><span>dragons</span><span> </span><span>&lt;-</span><span> </span><span>within</span><span>(</span><span>dragons</span><span>,</span><span> </span><span>sample</span><span> </span><span>&lt;-</span><span> </span><span>factor</span><span>(</span><span>mountainRange</span><span>:</span><span>site</span><span>))</span><span>
</span></code></pre></div></div>
<p>Now it’s obvious that we have 24 samples (8 mountain ranges x 3 sites) and not just 3: our <code>sample</code> is a 24-level factor and we should use that instead of using <code>site</code> in our models: each site belongs to a specific mountain range.</p>
<p><strong>To sum up:</strong> for <strong>nested random effects</strong>, the factor appears <strong>ONLY</strong> within a particular level of another factor (each site belongs to a specific mountain range and only to that range); for <strong>crossed effects</strong>, a given factor appears in more than one level of another factor (dragons appearing within more than one mountain range). <strong>Or you can just remember that if your random effects aren’t nested, then they are crossed!</strong></p>
<h3>Our second mixed model</h3>
<p>Based on the above, using following specification would be <strong><u>wrong</u></strong>:</p>
<div><div><pre><code><span>mixed.WRONG</span><span> </span><span>&lt;-</span><span> </span><span>lmer</span><span>(</span><span>testScore</span><span> </span><span>~</span><span> </span><span>bodyLength2</span><span> </span><span>+</span><span> </span><span>(</span><span>1</span><span>|</span><span>mountainRange</span><span>)</span><span> </span><span>+</span><span> </span><span>(</span><span>1</span><span>|</span><span>site</span><span>),</span><span> </span><span>data</span><span> </span><span>=</span><span> </span><span>dragons</span><span>)</span><span> </span><span># treats the two random effects as if they are crossed</span><span>
</span></code></pre></div></div>
<p>But we can go ahead and fit a new model, one that takes into account both the differences between the mountain ranges, as well as the differences between the sites within those mountain ranges by using our <code>sample</code> variable.</p>
<p>Our question gets <strong>adjusted slightly again</strong>: Is there an association between body length and intelligence in dragons <strong><em>after</em></strong> controlling for variation in mountain ranges and sites within mountain ranges?</p>
<div><div><pre><code><span>mixed.lmer2</span><span> </span><span>&lt;-</span><span> </span><span>lmer</span><span>(</span><span>testScore</span><span> </span><span>~</span><span> </span><span>bodyLength2</span><span> </span><span>+</span><span> </span><span>(</span><span>1</span><span>|</span><span>mountainRange</span><span>)</span><span> </span><span>+</span><span> </span><span>(</span><span>1</span><span>|</span><span>sample</span><span>),</span><span> </span><span>data</span><span> </span><span>=</span><span> </span><span>dragons</span><span>)</span><span> </span><span># the syntax stays the same, but now the nesting is taken into account</span><span>
</span><span>summary</span><span>(</span><span>mixed.lmer2</span><span>)</span><span>
</span></code></pre></div></div>
<p>Here, we are trying to account for <strong>all the mountain-range-level</strong> <em>and</em> <strong>all the site-level influences</strong> and we are hoping that our random effects have soaked up all these influences so we can control for them in the model.</p>
<p>For the record, you could also use the below syntax, but I’d advise you to set out your variables properly and make sure nesting is stated explicitly within them, that way you don’t have to remember to specify the nesting.</p>
<p><code>(1|mountainRange/site)</code> or even
<code>(1|mountainRange) + (1|mountainRange:site)</code></p>
<p>Let’s plot this again - visualising what’s going on is always helpful. You should be able to see eight mountain ranges with three sites (different colour points) within them, with a line fitted through each site.</p>
<div><div><pre><code><span>(</span><span>mm_plot</span><span> </span><span>&lt;-</span><span> </span><span>ggplot</span><span>(</span><span>dragons</span><span>,</span><span> </span><span>aes</span><span>(</span><span>x</span><span> </span><span>=</span><span> </span><span>bodyLength</span><span>,</span><span> </span><span>y</span><span> </span><span>=</span><span> </span><span>testScore</span><span>,</span><span> </span><span>colour</span><span> </span><span>=</span><span> </span><span>site</span><span>))</span><span> </span><span>+</span><span>
</span><span>facet_wrap</span><span>(</span><span>~</span><span>mountainRange</span><span>,</span><span> </span><span>nrow</span><span>=</span><span>3</span><span>)</span><span> </span><span>+</span><span>
</span><span>geom_point</span><span>()</span><span> </span><span>+</span><span>
</span><span>theme_classic</span><span>()</span><span> </span><span>+</span><span>
</span><span>geom_line</span><span>(</span><span>data</span><span> </span><span>=</span><span> </span><span>cbind</span><span>(</span><span>dragons</span><span>,</span><span> </span><span>pred</span><span> </span><span>=</span><span> </span><span>predict</span><span>(</span><span>mixed.lmer2</span><span>)),</span><span> </span><span>aes</span><span>(</span><span>y</span><span> </span><span>=</span><span> </span><span>pred</span><span>))</span><span> </span><span>+</span><span>
</span><span>theme</span><span>(</span><span>legend.position</span><span> </span><span>=</span><span> </span><span>"none"</span><span>))</span><span>
</span></code></pre></div></div>
<center><img alt="Img" src="https://ourcodingclub.github.io/img/mm-10.png"></center>
<p><strong>Well done for getting here!</strong> You have now fitted mixed models and you know how to account for crossed random effects too. You saw that failing to account for the correlation in data might lead to misleading results - it seemed that body length affected the test score until we accounted for the variation coming from mountain ranges. We can see now that body length doesn’t influence the test scores - great! We can pick smaller dragons for any future training - smaller ones should be more manageable! ;]</p>
<p>If you are particularly keen, the next section gives you a few options when it comes to <strong>presenting your model results</strong> and in the last “extra” section you can learn about the <strong>model selection conundrum</strong>. There is just a little bit more code there to get through if you fancy those.</p>
<h4>Presenting your model results</h4>
<p>Once you get your model, you have to <strong>present</strong> it in a nicer form.</p>
<h4>Tables</h4>
<p>For <code>lme4</code>, if you are looking for a table, I’d recommend that you have a look at the <code>stargazer</code> package.</p>
<p><code>stargazer</code>is very nicely annotated and there are lots of resources (e.g. <a href="https://cran.r-project.org/web/packages/stargazer/vignettes/stargazer.pdf" target="_blank">this</a>) out there and a <a href="http://jakeruss.com/cheatsheets/stargazer.html" target="_blank">great cheat sheet</a> so I won’t go into too much detail, as I’m confident you will find everything you need.</p>
<p>Here is a quick example - simply plug in your model name, in this case <code>mixed.lmer2</code> into the <code>stargazer</code> function. I set <code>type</code> to <code>"text"</code> so that you can see the table in your console. I usually tweak the table like this until I’m happy with it and then export it using <code>type = "latex"</code>, but <code>"html"</code> might be more useful for you if you are not a LaTeX user.</p>
<p>If you are keen, explore this table a little further - what would you change? What would you get rid off?</p>
<div><div><pre><code><span>stargazer</span><span>(</span><span>mixed.lmer2</span><span>,</span><span> </span><span>type</span><span> </span><span>=</span><span> </span><span>"text"</span><span>,</span><span>
</span><span>digits</span><span> </span><span>=</span><span> </span><span>3</span><span>,</span><span>
</span><span>star.cutoffs</span><span> </span><span>=</span><span> </span><span>c</span><span>(</span><span>0.05</span><span>,</span><span> </span><span>0.01</span><span>,</span><span> </span><span>0.001</span><span>),</span><span>
</span><span>digit.separator</span><span> </span><span>=</span><span> </span><span>""</span><span>)</span><span>
</span></code></pre></div></div>
<center><img alt="Img" src="https://ourcodingclub.github.io/img/mm-tab.png"></center>
<p>If you are looking for <strong>a way to create plots of your results</strong> check out <code>dotwhisker</code> and this <a href="https://cran.r-project.org/web/packages/dotwhisker/vignettes/dotwhisker-vignette.html" target="_blank">tutorial</a>.</p>
<h4>Further processing</h4>
<p>If you’d like to be able <strong>to do more with your model results</strong>, for instance process them further, collate model results from multiple models or plot, them have a look at the <code>broom</code> package. This <a href="http://varianceexplained.org/r/broom-intro/" target="_blank">tutorial</a> is a great start.</p>
<h4>EXTRA: P-values and model selection</h4>
<p>Please be <strong>very, very careful</strong> when it comes to model selection. Focus on your <strong>question</strong>, don’t just plug in and drop variables from a model haphazardly until you make something “significant”. Always choose variables based on biology/ecology: I might use model selection to check a couple of non-focal parameters, but I keep the “core” of the model untouched in most cases. <strong>Define your goals and questions and focus on that.</strong> Also, don’t just put all possible variables in (i.e. don’t <strong>overfit</strong>). Remember that as a rule of thumb, <strong>you need 10 times more data than parameters</strong> you are trying to estimate.</p>
<p>For more info on overfitting check out this <a href="https://ourcodingclub.github.io/2017/02/28/modelling.html" target="_blank">tutorial</a>.</p>
<h4>Fixed effects structure</h4>
<p><strong>Before we start, again: think twice before trusting model selection!</strong></p>
<p>Most of you are probably going to be predominantly interested in your fixed effects, so let’s start here. <code>lme4</code> doesn’t spit out p-values for the parameters by default. This is a conscious choice made by the authors of the package, as there are many problems with p-values (I’m sure you are aware of the debates!).</p>
<p>You will inevitably look for a way to assess your model though so here are a few solutions on how to go about hypothesis testing in linear mixed models (LMMs):</p>
<p><strong>From worst to best:</strong></p>
<ul>
<li>Wald Z-tests</li>
<li>Wald t-tests (but LMMs need to be balanced and nested)</li>
<li>Likelihood ratio tests (via <code>anova()</code> or <code>drop1()</code>)</li>
<li>
<code>MCMC</code> or parametric bootstrap confidence intervals</li>
</ul>
<p>See <a href="http://stats.stackexchange.com/questions/95054/how-to-get-an-overall-p-value-and-effect-size-for-a-categorical-factor-in-a-mi" target="_blank">this link</a> for more information and further reading.</p>
<p>I think that <code>MCMC</code> and bootstrapping are a bit out of our reach for this workshop so let’s have a quick go at <strong>likelihood ratio tests</strong> using <code>anova()</code>. With large sample sizes, p-values based on the likelihood ratio are generally considered okay. <strong>NOTE:</strong> With small sample sizes, you might want to look into deriving p-values using the Kenward-Roger or Satterthwaite approximations (for <code>REML</code> models). Check out the <code>pbkrtest</code> package.</p>
<p>Fit the models, a full model and a reduced model in which we dropped our fixed effect (<code>bodyLength2</code>):</p>
<div><div><pre><code><span>full.lmer</span><span> </span><span>&lt;-</span><span> </span><span>lmer</span><span>(</span><span>testScore</span><span> </span><span>~</span><span> </span><span>bodyLength2</span><span> </span><span>+</span><span> </span><span>(</span><span>1</span><span>|</span><span>mountainRange</span><span>)</span><span> </span><span>+</span><span> </span><span>(</span><span>1</span><span>|</span><span>sample</span><span>),</span><span> </span><span>data</span><span> </span><span>=</span><span> </span><span>dragons</span><span>,</span><span> </span><span>REML</span><span> </span><span>=</span><span> </span><span>FALSE</span><span>)</span><span>
</span><span>reduced.lmer</span><span> </span><span>&lt;-</span><span> </span><span>lmer</span><span>(</span><span>testScore</span><span> </span><span>~</span><span> </span><span>1</span><span> </span><span>+</span><span> </span><span>(</span><span>1</span><span>|</span><span>mountainRange</span><span>)</span><span> </span><span>+</span><span> </span><span>(</span><span>1</span><span>|</span><span>sample</span><span>),</span><span> </span><span>data</span><span> </span><span>=</span><span> </span><span>dragons</span><span>,</span><span> </span><span>REML</span><span> </span><span>=</span><span> </span><span>FALSE</span><span>)</span><span>
</span></code></pre></div></div>
<p>Compare them:</p>
<div><div><pre><code><span>anova</span><span>(</span><span>reduced.lmer</span><span>,</span><span> </span><span>full.lmer</span><span>)</span><span> </span><span># the two models are not significantly different</span><span>
</span></code></pre></div></div>
<p>Notice that we have fitted our models with <code>REML = FALSE</code>.</p>
<p><strong>REML</strong> stands for <strong>restricted (or “residual”) maximum likelihood</strong> and it is the default parameter estimation criterion for linear mixed models. As you probably guessed, <strong>ML</strong> stands for <strong>maximum likelihood</strong> - you can set <code>REML = FALSE</code> in your call to <code>lmer</code> to use ML estimates. However, <strong>ML estimates are known to be biased</strong> and with REML being usually less biased, <strong>REML estimates of variance components are generally preferred.</strong> This is why in our previous models we skipped setting <code>REML</code> - we just left it as default (i.e. <code>REML = TRUE</code>).</p>
<p><strong>REML</strong> assumes that the fixed effects structure is correct. You <strong>should use maximum likelihood when comparing models with different fixed effects</strong>, as <strong>ML</strong> doesn’t rely on the coefficients of the fixed effects - and that’s why we are refitting our full and reduced models above with the addition of <code>REML = FALSE</code> in the call.</p>
<p>Even though you <strong>use ML to compare models</strong>, you should <strong>report parameter estimates from your final “best” REML model</strong>, as ML may underestimate variance of the random effects.</p>
<p><strong>NOTE 2:</strong> Models can also be compared using the <code>AICc</code> function from the <code>AICcmodavg</code> package. The Akaike Information Criterion (AIC) is a measure of model quality. AICc corrects for bias created by small sample size when estimating AIC. Generally, if models are within 2 AICc units of each other they are very similar. Within 5 units they are quite similar, over 10 units difference and you can probably be happy with the model with lower AICc. As with p-values though, there is no “hard line” that’s always correct.</p>
<p><strong>NOTE 3:</strong> There isn’t really an agreed upon way of dealing with the variance from the random effects in mixed models when it comes to assessing significance. Both <strong>p-values</strong> and <strong>effect sizes</strong> have issues, although from what I gather, p-values seem to cause more disagreement than effect sizes, at least in the R community.</p>
<h4>Random effects structure</h4>
<p>Now you might wonder about selecting your random effects. In general, I’d advise you to think about your <strong>experimental design, your system and data collected, as well as your questions</strong>.</p>
<p>If your random effects are there to deal with <strong>pseudoreplication</strong>, then it doesn’t really matter whether they are “significant” or not: they <strong>are part of your design</strong> and have to be included. Imagine we tested our dragons multiple times - we then <em>have to</em> fit dragon identity as a random effect.</p>
<p>On the other hand, if you are trying to account for other variability that you think might be important, it becomes a bit harder. Imagine we measured the mass of our dragons over their lifespans (let’s say 100 years). We might then want to fit year as a random effect to account for any temporal variation - maybe some years were affected by drought, the resources were scarce and so dragon mass was negatively impacted. Year would definitely be a sensible random effect, although strictly speaking not a must.</p>
<p>When it comes to such random effects you can use <strong>model selection</strong> to help you decide what to keep in. Following Zuur’s advice, we <strong>use <code>REML</code> estimators for comparison of models with different random effects</strong> (we keep fixed effects constant). (Zuur: “Two models with nested random structures cannot be done with ML because the estimators for the variance terms are biased.” )</p>
<p><strong>NOTE:</strong> Do <strong>NOT</strong> vary random and fixed effects at the same time - either deal with your random effects structure or with your fixed effects structure at any given point.</p>
<p><strong>NOTE 2:</strong> Do <strong>NOT</strong> compare <code>lmer</code> models with <code>lm</code> models (or <code>glmer</code> with <code>glm</code>).</p>
<h4>Entire model selection</h4>
<p>A few notes on the process of model selection. There are two ways here: (i) <strong>“top-down”</strong>, where you start with a complex model and gradually reduce it, and (ii) <strong>“step up”</strong>, where you start with a simple model and add new variables to it. Unfortunately, you might arrive at different final models by using those strategies and so you need to be careful.</p>
<p>The model selection process recommended by Zuur <em>et al.</em> (2009) is a top-down strategy and goes as follows:</p>
<ol>
<li>fit a <strong>full model</strong> (he even recommends “beyond optimal” i.e. more complex than you’d expect or want it to be)</li>
<li>sort out the <strong>random effects structure</strong> (use <code>REML</code> likelihoods or <code>REML</code> AIC or BIC)</li>
<li>sort out <strong>fixed effects structure</strong> (either use <code>REML</code> the F-statistic or the t-statistic or compare nested <code>ML</code> models - keep your random effects constant)</li>
<li>once you arrive at the <strong>final model present it using <code>REML</code> estimation</strong>
</li>
</ol>
<p><strong>NOTE:</strong> At the risk of sounding like a broken record: I think it’s best to decide on what your model is based on biology/ecology/data structure <em>etc</em>. than through following model selection blindly. Additionally, just because something is non-significant doesn’t necessarily mean you should always get rid of it.</p>
<h3>THE END</h3>
<p><strong>Well done for getting through this!</strong> As you probably gather, mixed effects models can be a bit tricky and often there isn’t much consensus on the best way to tackle something within them. The coding bit is actually the (relatively) easy part here. Be mindful of what you are doing, prepare the data well and things should be alright.</p>
<hr>
<hr>
<p><strong>Check out <a href="https://ourcodingclub.github.io/workshop/" target="_blank">this page</a> to learn how you can get involved! We are very happy to have people use our tutorials and adapt them to their needs. We are also very keen to expand the content on the website, so feel free to get in touch if you’d like to write a tutorial!</strong></p>
<p>This work is licensed under a <a href="https://creativecommons.org/licenses/by-sa/4.0/">Creative Commons Attribution-ShareAlike 4.0 International License</a>. <a href="https://creativecommons.org/licenses/by-sa/4.0/"><img alt="Img" src="https://licensebuttons.net/l/by-sa/4.0/80x15.png"></a></p>
<h3><a href="https://www.surveymonkey.co.uk/r/HJYGVSF" target="_blank">  We would love to hear your feedback, please fill out our survey!</a></h3>
<h3>  You can contact us with any questions on <a href="mailto:ourcodingclub@gmail.com?Subject=Tutorial%20question" target="_top">ourcodingclub@gmail.com</a>
</h3>
<h3>  Related tutorials:</h3>
<ul>
<li>
<a href="http://ourcodingclub.github.io/2018/11/26/earth-engine.html">
  - Intro to the Google Earth Engine
</a>
</li>
<li>
<a href="http://ourcodingclub.github.io/2018/10/15/machine-learning.html">
  - Intro to Machine Learning in R (K Nearest Neighbours Algorithm)
</a>
</li>
<li>
<a href="http://ourcodingclub.github.io/2018/04/30/stan-2.html">
  - Generalised linear models in Stan
</a>
</li>
<li>
<a href="http://ourcodingclub.github.io/2018/04/17/stan-intro.html">
  - Intro to Stan
</a>
</li>
<li>
<a href="http://ourcodingclub.github.io/2018/04/06/model-design.html">
  - Intro to model design
</a>
</li>
<li>
<a href="http://ourcodingclub.github.io/2018/01/29/qualitative.html">
  - Analysing ordinal data, surveys, count data
</a>
</li>
<li>
<a href="http://ourcodingclub.github.io/2018/01/22/mcmcglmm.html">
  - Meta-analysis for biologists using MCMCglmm
</a>
</li>
<li>
<a href="http://ourcodingclub.github.io/2017/04/26/time.html">
  - Analysing Time Series Data
</a>
</li>
</ul>
<h3>  Subscribe to our mailing list:</h3>
</div>
</div>
</div>
</div>
</section>
</div></div><br>
<br>
<br>
<br>
<br>
<br>
via ourcodingclub.github.io http://bit.ly/2uAMVVw<br>
<br>
May 4, 2019 at 10:33PM
